### ISSUE TITLE
When i run paper experiment llama2-7B，I have encountered an unsolvable problem。

### ISSUE THREAD

**Issue by Destiny-Hao** [Ledzy/BAdam#13]
![1730808385294](https://github.com/user-attachments/assets/caa0fc4b-9ad9-465f-aef9-98b48ad1dafc)
Can you give me some suggestions to solve this problem? Thank you！

**Comment by Ledzy** [comment]
Hi, this seems to be related to the version of the library `datasets`. Please try downgrading it, i.e. `pip install datasets==2.15.0`.

**Comment by Destiny-Hao** [comment]
> Hi, this seems to be related to the version of the library `datasets`. Please try downgrading it, i.e. `pip install datasets==2.15.0`.

Thank you, this method is very useful！I have encountered another problem now. Can you give me some advice? Thank you！
![1730855619611](https://github.com/user-attachments/assets/b90aeb4f-6eb0-4511-a1c8-4426493d8992)

**Comment by Ledzy** [comment]
Hi, you may check if the `pure_bf16` flat is set in the running command. If so, remove it, as BAdam use mixed precision training. To reproduce our experiment results in training Llama 2-7B, you may follow the instruction in `readme.md` and use the provided command.

**Comment by Destiny-Hao** [comment]
> Hi, you may check if the `pure_bf16` flat is set in the running command. If so, remove it, as BAdam use mixed precision training. To reproduce our experiment results in training Llama 2-7B, you may follow the instruction in `readme.md` and use the provided command.

CUDA_VISIBLE_DEVICES="0, 1, 2, 3" python src/train_bash.py \
    --stage sft \
    --model_name_or_path /home/mh/llama2-7b \
    --do_train \
    --dataset commonsense-170k \
    --template default \
    --finetuning_type block \
    --output_dir ./outputs/llama2-7b \
    --overwrite_cache \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --lr_scheduler_type cosine \
    --logging_steps 1 \
    --save_steps 1000 \
    --val_size 500 \
    --eval_steps 20 \
    --evaluation_strategy steps \
    --learning_rate 1e-6 \
    --num_train_epochs 3 \
    --overwrite_output_dir \
    --plot_loss \
    --switch_block_every 100 \
    --switch_mode random
This is the command I used, only changing the model path and dataset path, I cannot find pure-bf16.

**Comment by Ledzy** [comment]
Hi, could you use the same command and see if the code works? You may want to set CUDA_VISIBLE_DEVICES=0 since the code is based on single GPU. 

Please be noted that the code for Llama 2-7B in this repo is purely for the purpose of reproducing our paper's experimental results on Alpaca-GPT4 dataset. If you just want to train Llama 2-7B on your dataset, you may use [Llama Factory](https://github.com/hiyouga/LLaMA-Factory), which has intergrated BAdam and other up-to-date tools of LLM training.

**Comment by Destiny-Hao** [comment]
Hello, I used your code to perform supervised fine-tuning of llama2-7b, and then wrote my own test code. I found that in your code, the model weight file saved at each checkpoint seems to only have one layer. If that's the case, can you provide me with some methods to test the fine tuned model? thank you！




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Ledzy/BAdam"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年11月6日(星期三) 中午12:05
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Ledzy/BAdam] When i run paper experiment llama2-7B，I have encountered an unsolvable problem。 (Issue #13)





 
Hi, could you use the same command and see if the code works? You may want to set CUDA_VISIBLE_DEVICES=0 since the code is based on single GPU.
 
Please be noted that the code for Llama 2-7B in this repo is purely for the purpose of reproducing our paper's experimental results on Alpaca-GPT4 dataset. If you just want to train Llama 2-7B on your dataset, you may use Llama Factory, which has intergrated BAdam and other up-to-date tools of LLM training.

**Comment by Ledzy** [comment]
To save the whole model instead of a single layer, change the following line in `llama-alpaca/src/llmtuner/tuner/core/trainer.py`:
* line 61: `if self.finetuning_args.finetuning_type in ["full", "block"]:`
* line 82: `if self.finetuning_args.finetuning_type in ["full", "block"] and self.tokenizer is not None:`

**Comment by Destiny-Hao** [comment]
Thank you very much!



---Original---
From: ***@***.***&gt;
Date: Fri, Nov 8, 2024 17:47 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Ledzy/BAdam] When i run paper experiment llama2-7B，I have encountered an unsolvable problem。 (Issue #13)




 
To save the whole model instead of a single layer, change the following line in llama-alpaca/src/llmtuner/tuner/core/trainer.py:
  
line 61: if self.finetuning_args.finetuning_type in ["full", "block"]:
 
line 82: if self.finetuning_args.finetuning_type in ["full", "block"] and self.tokenizer is not None:

**Comment by Destiny-Hao** [comment]
Hello, I'm sorry to bother you again. The trained model outputs garbled code during testing. Can you give me some suggestions to solve this problem?
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Ledzy/BAdam"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年11月11日(星期一) 中午1:02
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Ledzy/BAdam] When i run paper experiment llama2-7B，I have encountered an unsolvable problem。 (Issue #13)





 
Closed #13 as completed.

**Comment by Ledzy** [comment]
You may check the loss first, then check if the template is correctly set.
